Ability to use IndexTriangle

w0rm opened the pull request:
This is my work on top of @nacmartin [pull request](https://github.com/elm-community/elm-webgl/pull/40). I only made `IndexTriangle` to take a list of tuples of indices (similar to `Triangle`), removed unnecessary tuple wrapper, and modified one example to use this feature.

This PR closes [this issue](https://github.com/elm-community/elm-webgl/issues/37) by @7sharp9.

However I am not sure about a few things:

1. having to reverse `bufferElems` in `doBindAttribute`, I couldn't find enough explanation
2. introducing indexes only for `Triangle`, and not other drawables. We may have to reconsider the API for Drawables, to be able to create either indexed or non-indexed drawables.
3. duplicating some functionality in the native module (`doBindSetup` and `doBindSetupIndexed` are almost identical, except that the the first one generates indices and the second takes them from the input)

nacmartin committed:
Ability fo use IndexTriangles

Reverse elements before filling the buffer

The elements were in reversed order, and this messes up when working with indices.

nacmartin commented:
About the reverse, I don't have a good explanation, but it looks like if you do `A2(listMap,` items are accessed in the reverse order that you would expect them to be. This is not a problem if you are not using indexes, since everything goes, but it is a problem if you care about the order to refer to them in the indexes.

w0rm commented:
@nacmartin I see, thanks. Will investigate how to avoid unnecessary iteration here.

w0rm commented:
@nacmartin I pulled the common part of indices into a separate function. `List.map` is always reversing the order because it needs to build up a new list. We could use `Native.List.toArray` from the Core, but I'm not sure if we can rely on it, so I'd rather keep `reverse` in there.

w0rm commented:
I have another concern, what if an index points to a non-existing attribute, that is outside the list bounds?

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -226,10 +229,10 @@ var _elm_community$webgl$Native_WebGL = function () {
       dataFill(array, attributeInfo.size, dataIdx, elem, attribute.name);
       dataIdx += attributeInfo.size * elemSize;
-    }, bufferElems);
+    }), undefined, bufferElems);
```
:scream:

eeue56 replied:
Why undefined?

eeue56 replied:
it would be better to just do this by hand rather than using list methods in an impure way.

w0rm replied:
But what if internal representation of List changes?

w0rm replied:
@eeue56 I think this is totally fine,  this is discussed here https://github.com/elm-community/webgl/issues/12

eeue56 replied:
The internal representation of List hasn't changed for years. Other libraries, such as the virtual-dom, also use their own function for unwrapping lists.

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -242,46 +245,79 @@ var _elm_community$webgl$Native_WebGL = function () {
+  *  @param {Number} numIndices The number of indices
+  *  @return {Uint16Array} indices
+  */
+  function doBindSetupSequential(gl, numIndices) {
```
this doesn't need to take gl

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -242,46 +245,79 @@ var _elm_community$webgl$Native_WebGL = function () {
+  *  @param {Number} indexSize the size of the index
+  *  @return {Uint16Array} indices
+  */
+  function doBindSetupIndexed(gl, indicesList, indexSize) {
```
this doesn't need to take gl

w0rm replied:
Oh, totally forgot to get rid of it here, thanks

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -242,46 +245,79 @@ var _elm_community$webgl$Native_WebGL = function () {
+        indices[fillOffset++] = elem;
+      } else {
+        for (i = 0; i < indexSize; i++) {
+          indices[fillOffset++] = elem['_' + i];
```
fillOffset++ i better written as `fillOffset + i`

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -242,46 +245,79 @@ var _elm_community$webgl$Native_WebGL = function () {
+          indices[fillOffset++] = elem['_' + i];
+        }
+      }
+    }), undefined, indicesList);
```
don't use core's foldL here. just do a simple map

w0rm replied:
Map builds a result that is ignored and iterates backwards

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -242,46 +245,79 @@ var _elm_community$webgl$Native_WebGL = function () {
+        indices[fillOffset++] = elem;
+      } else {
+        for (i = 0; i < indexSize; i++) {
+          indices[fillOffset++] = elem['_' + i];
```
don't trust converstion - use `i.toString()`

w0rm replied:
This will never break though

eeue56 replied:
Explicit is better than implicit, especially when it comes to javascript

w0rm replied:
@eeue56 sure thing, I've fixed this

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -382,14 +416,14 @@ var _elm_community$webgl$Native_WebGL = function () {
+        A3(listFoldl, F2(function (functionCall) {
           functionCall(gl);
-        }, render.functionCalls);
+        }), undefined, render.functionCalls);
```
don't use core's foldl here

w0rm replied:
@eeue56 function calls should be applied in the original order and not reversed

eeue56 commented on src/Native/WebGL.js:
```diff
@@ -581,9 +615,9 @@ var _elm_community$webgl$Native_WebGL = function () {
 
     if (gl) {
-      A2(listMap, function (functionCall) {
+      A3(listFoldl, F2(function (functionCall) {
```
don't use core's foldl here

w0rm replied:
@eeue56 but why not?

eeue56 commented:
I took a look through, and there's a bunch of places where you would be better off writing a function to iterate through the list as you intend yourself. You don't get much benefit from using core's foldl if you're using it in an impure way, and it would be simpler to follow without it

w0rm commented:
@eeue56 cool, thanks! I addressed some of your concerns (put some irrelevant changes back), can we continue the discussion about lists in this thread https://github.com/elm-community/webgl/issues/12 ?

Commit:
Iterate over internal list representations

w0rm commented:
@eeue56 so I've implemented native list iterations, and decided to not use anything from the `List` module. This changes the order of drawables, but this PR bumps a major version anyway, and things that depend on this may be easily fixed with an extra `List.reverse`.

Zinggi commented:
@w0rm maybe you can add [this link](http://www.opengl-tutorial.org/intermediate-tutorials/tutorial-9-vbo-indexing/) to the documentation for `IndexedTriangle`?
It has a nice image to make the concept clear.
